File: testers/sqlite/scala/src/scala/testdb/SqlScriptTest.scala
package testdb

import dev.typr.foundationssc.{Connection, ConnectionRead}
import org.junit.Assert.*
import org.junit.Test
import testdb.all_scalar_types_search.AllScalarTypesSearchSqlRepoImpl
import testdb.customer_search.CustomerSearchSqlRepoImpl
import testdb.delete_old_orders.DeleteOldOrdersSqlRepoImpl
import testdb.order_summary_by_customer.OrderSummaryByCustomerSqlRepoImpl
import testdb.product_details_with_sales.ProductDetailsWithSalesSqlRepoImpl

import java.time.{LocalDate, LocalDateTime}

class SqlScriptTest {
  private val customerSearch = new CustomerSearchSqlRepoImpl()
  private val orderSummary = new OrderSummaryByCustomerSqlRepoImpl()
  private val productDetails = new ProductDetailsWithSalesSqlRepoImpl()
  private val scalarSearch = new AllScalarTypesSearchSqlRepoImpl()
  private val deleteOld = new DeleteOldOrdersSqlRepoImpl()

  @Test def customerSearchAll(): Unit = withConnection {
    val rows = customerSearch(None, None, None, 100L)(using summon[ConnectionRead])
    assertEquals(2, rows.size)
  }

  @Test def customerSearchByName(): Unit = withConnection {
    val rows = customerSearch(Some("John%"), None, None, 100L)(using summon[ConnectionRead])
    assertEquals(1, rows.size)
    assertEquals("John Doe", rows.head.name)
  }

  @Test def customerSearchByEmail(): Unit = withConnection {
    val rows = customerSearch(None, Some("%jane%"), None, 100L)(using summon[ConnectionRead])
    assertEquals(1, rows.size)
  }

  @Test def customerSearchLimit(): Unit = withConnection {
    val rows = customerSearch(None, None, None, 1L)(using summon[ConnectionRead])
    assertEquals(1, rows.size)
  }

  @Test def customerSearchByCreatedAfter(): Unit = withConnection {
    val rows = customerSearch(None, None, Some(LocalDateTime.of(2099, 1, 1, 0, 0)), 100L)(using summon[ConnectionRead])
    assertEquals(0, rows.size)
  }

  @Test def orderSummaryInRange(): Unit = withConnection {
    val rows = orderSummary(LocalDate.of(2025, 1, 1), LocalDate.of(2025, 12, 31))(using summon[ConnectionRead])
    assertEquals(2, rows.size)
    val total = rows.flatMap(_.totalRevenue).sum
    assertEquals(0, BigDecimal("109.97").compare(total))
  }

  @Test def orderSummaryEmptyRange(): Unit = withConnection {
    val rows = orderSummary(LocalDate.of(2099, 1, 1), LocalDate.of(2099, 12, 31))(using summon[ConnectionRead])
    assertEquals(0, rows.size)
  }

  @Test def productDetailsAll(): Unit = withConnection {
    val rows = productDetails(None)(using summon[ConnectionRead])
    assertEquals(2, rows.size)
  }

  @Test def productDetailsMinPrice(): Unit = withConnection {
    val rows = productDetails(Some(BigDecimal("40.00")))(using summon[ConnectionRead])
    assertEquals(1, rows.size)
    assertEquals("Widget B", rows.head.productName)
  }

  @Test def allScalarSearchAll(): Unit = withConnection {
    val rows = scalarSearch(None, None, None)(using summon[ConnectionRead])
    assertEquals(1, rows.size)
  }

  @Test def allScalarSearchMinId(): Unit = withConnection {
    val rows = scalarSearch(Some(100L), None, None)(using summon[ConnectionRead])
    assertEquals(0, rows.size)
  }

  @Test def allScalarSearchByText(): Unit = withConnection {
    val rows = scalarSearch(None, Some("hel%"), None)(using summon[ConnectionRead])
    assertEquals(1, rows.size)
  }

  @Test def deleteOldOrders(): Unit = withConnection {
    val ordersRepo = new testdb.orders.OrdersRepoImpl()
    ordersRepo.insert(
      testdb.orders.OrdersRow(
        testdb.orders.OrdersId(9001L),
        testdb.customers.CustomersId(1L),
        LocalDate.of(2020, 1, 1),
        Some(BigDecimal("10.00")),
        "completed"
      )
    )
    val deleted = deleteOld(LocalDate.of(2025, 1, 1))
    assertEquals(1, deleted)
  }

  @Test def deleteOldOrdersNoneMatch(): Unit = withConnection {
    val deleted = deleteOld(LocalDate.of(1900, 1, 1))
    assertEquals(0, deleted)
  }
}

File: testers/sqlite/kotlin/src/kotlin/testdb/SqlScriptTest.kt
package testdb

import org.junit.Assert.*
import org.junit.Test
import testdb.all_scalar_types_search.AllScalarTypesSearchSqlRepoImpl
import testdb.customer_search.CustomerSearchSqlRepoImpl
import testdb.delete_old_orders.DeleteOldOrdersSqlRepoImpl
import testdb.order_summary_by_customer.OrderSummaryByCustomerSqlRepoImpl
import testdb.product_details_with_sales.ProductDetailsWithSalesSqlRepoImpl
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime

class SqlScriptTest {
    private val customerSearch = CustomerSearchSqlRepoImpl()
    private val orderSummary = OrderSummaryByCustomerSqlRepoImpl()
    private val productDetails = ProductDetailsWithSalesSqlRepoImpl()
    private val scalarSearch = AllScalarTypesSearchSqlRepoImpl()
    private val deleteOld = DeleteOldOrdersSqlRepoImpl()

    @Test
    fun customerSearchAll() {
        SqliteTestHelper.applyRead { c ->
            val rows = customerSearch.apply(null, null, null, 100L, c)
            assertEquals(2, rows.size)
        }
    }

    @Test
    fun customerSearchByName() {
        SqliteTestHelper.applyRead { c ->
            val rows = customerSearch.apply("John%", null, null, 100L, c)
            assertEquals(1, rows.size)
            assertEquals("John Doe", rows[0].name)
        }
    }

    @Test
    fun customerSearchByEmail() {
        SqliteTestHelper.applyRead { c ->
            val rows = customerSearch.apply(null, "%jane%", null, 100L, c)
            assertEquals(1, rows.size)
        }
    }

    @Test
    fun customerSearchLimit() {
        SqliteTestHelper.applyRead { c ->
            val rows = customerSearch.apply(null, null, null, 1L, c)
            assertEquals(1, rows.size)
        }
    }

    @Test
    fun customerSearchByCreatedAfter() {
        SqliteTestHelper.applyRead { c ->
            val rows = customerSearch.apply(null, null, LocalDateTime.of(2099, 1, 1, 0, 0), 100L, c)
            assertEquals(0, rows.size)
        }
    }

    @Test
    fun orderSummaryInRange() {
        SqliteTestHelper.applyRead { c ->
            val rows = orderSummary.apply(LocalDate.of(2025, 1, 1), LocalDate.of(2025, 12, 31), c)
            assertEquals(2, rows.size)
            val total = rows.mapNotNull { it.totalRevenue }.fold(BigDecimal.ZERO) { a, b -> a.add(b) }
            assertEquals(0, BigDecimal("109.97").compareTo(total))
        }
    }

    @Test
    fun orderSummaryEmptyRange() {
        SqliteTestHelper.applyRead { c ->
            val rows = orderSummary.apply(LocalDate.of(2099, 1, 1), LocalDate.of(2099, 12, 31), c)
            assertEquals(0, rows.size)
        }
    }

    @Test
    fun productDetailsAll() {
        SqliteTestHelper.applyRead { c ->
            val rows = productDetails.apply(null, c)
            assertEquals(2, rows.size)
        }
    }

    @Test
    fun productDetailsMinPrice() {
        SqliteTestHelper.applyRead { c ->
            val rows = productDetails.apply(BigDecimal("40.00"), c)
            assertEquals(1, rows.size)
            assertEquals("Widget B", rows[0].productName)
        }
    }

    @Test
    fun allScalarSearchAll() {
        SqliteTestHelper.applyRead { c ->
            val rows = scalarSearch.apply(null, null, null, c)
            assertEquals(1, rows.size)
        }
    }

    @Test
    fun allScalarSearchMinId() {
        SqliteTestHelper.applyRead { c ->
            val rows = scalarSearch.apply(100L, null, null, c)
            assertEquals(0, rows.size)
        }
    }

    @Test
    fun allScalarSearchByText() {
        SqliteTestHelper.applyRead { c ->
            val rows = scalarSearch.apply(null, "hel%", null, c)
            assertEquals(1, rows.size)
        }
    }

    @Test
    fun deleteOldOrders() {
        SqliteTestHelper.run { c ->
            val ordersRepo = testdb.orders.OrdersRepoImpl()
            ordersRepo.insert(
                testdb.orders.OrdersRow(
                    testdb.orders.OrdersId(9001L),
                    testdb.customers.CustomersId(1L),
                    LocalDate.of(2020, 1, 1),
                    BigDecimal("10.00"),
                    "completed"
                ), c
            )
            val deleted = deleteOld.apply(LocalDate.of(2025, 1, 1), c)
            assertEquals(1, deleted)
        }
    }

    @Test
    fun deleteOldOrdersNoneMatch() {
        SqliteTestHelper.run { c ->
            val deleted = deleteOld.apply(LocalDate.of(1900, 1, 1), c)
            assertEquals(0, deleted)
        }
    }
}
